File: Editorial/148_Sort_List/JavaScript/2_Bottom_Up_Merge_Sort.js
var sortList = function (head) {
    if (!head || !head.next) return head;

    function getCount(head) {
        let cnt = 0,
            ptr = head;
        while (ptr) {
            ptr = ptr.next;
            cnt++;
        }
        return cnt;
    }

    function split(start, size) {
        let midPrev = start;
        let end = start.next;
        // use fast and slow approach to find middle and end of second linked list
        for (
            let index = 1;
            index < size && (midPrev.next || end.next);
            index++
        ) {
            if (end.next) {
                end = end.next.next ? end.next.next : end.next;
            }
            if (midPrev.next) {
                midPrev = midPrev.next;
            }
        }
        const mid = midPrev.next;
        midPrev.next = null;
        nextSubList = end.next;
        end.next = null;
        // return the start of second linked list
        return mid;
    }

    function merge(list1, list2) {
        let dummyHead = new ListNode();
        let newTail = dummyHead;
        while (list1 && list2) {
            if (list1.val < list2.val) {
                newTail.next = list1;
                list1 = list1.next;
            } else {
                newTail.next = list2;
                list2 = list2.next;
            }
            newTail = newTail.next;
        }
        newTail.next = list1 || list2;
        // traverse till the end of merged list to get the newTail
        while (newTail.next) {
            newTail = newTail.next;
        }
        // link the old tail with the head of merged list
        tail.next = dummyHead.next;
        // update the old tail to the new tail of merged list
        tail = newTail;
    }

    let n = getCount(head);
    let start = head;
    let dummyHead = new ListNode();
    let tail, nextSubList;

    for (let size = 1; size < n; size *= 2) {
        tail = dummyHead;
        while (start) {
            if (!start.next) {
                tail.next = start;
                break;
            }
            let mid = split(start, size);
            merge(start, mid);
            start = nextSubList;
        }
        start = dummyHead.next;
    }
    return dummyHead.next;
};

File: Editorial/148_Sort_List/Go/2_Bottom_Up_Merge_Sort.go
var (
    tail        *ListNode
    nextSubList *ListNode
)

func sortList(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
        return head
    }
    n := getCount(head)
    start := head
    dummyHead := &ListNode{}
    for size := 1; size < n; size = size * 2 {
        tail = dummyHead
        for start != nil {
            if start.Next == nil {
                tail.Next = start
                break
            }
            mid := split(start, size)
            merge(start, mid)
            start = nextSubList
        }
        start = dummyHead.Next
    }
    return dummyHead.Next
}

func split(start *ListNode, size int) *ListNode {
    midPrev := start
    end := start.Next
    for index := 1; index < size && (midPrev.Next != nil || end.Next != nil); index++ {
        if end.Next != nil {
            if end.Next.Next != nil {
                end = end.Next.Next
            } else {
                end = end.Next
            }
        }
        if midPrev.Next != nil {
            midPrev = midPrev.Next
        }
    }
    mid := midPrev.Next
    nextSubList = end.Next
    midPrev.Next = nil
    if end != nil {
        end.Next = nil
    }
    return mid
}

func merge(list1 *ListNode, list2 *ListNode) {
    dummyHead := &ListNode{}
    newTail := dummyHead
    for list1 != nil && list2 != nil {
        if list1.Val < list2.Val {
            newTail.Next = list1
            list1 = list1.Next
        } else {
            newTail.Next = list2
            list2 = list2.Next
        }
        newTail = newTail.Next
    }
    if list1 != nil {
        newTail.Next = list1
    } else {
        newTail.Next = list2
    }
    for newTail.Next != nil {
        newTail = newTail.Next
    }
    tail.Next = dummyHead.Next
    tail = newTail
}

func getCount(head *ListNode) int {
    cnt := 0
    ptr := head
    for ptr != nil {
        ptr = ptr.Next
        cnt++
    }
    return cnt
}
